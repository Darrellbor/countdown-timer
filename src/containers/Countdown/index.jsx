import React, { Component } from "react";

//components
import Header from "../../components/Header";
import Footer from "../../components/Footer";
import Controls from "../../components/Controls";
import SpeedControls from "../../components/SpeedControls";
import Alert from "../../components/Alert";
import Counter from "../../components/Counter";

//icons
import Pause from "../../components/Icons/pause";
import Play from "../../components/Icons/play";
import Stop from "../../components/Icons/stop";

//sound
import CountdownOverSound from "../../assets/sounds/countdown-complete.mp3";

class Countdown extends Component {
  state = {
    minute: "00",
    second: "00",
    halfMinute: "00",
    halfSecond: "00",
    clicked: "None",
    speedClicked: "None",
    activeControl: "None",
    activeSpeed: "1X",
    isAlertOpen: false,
    alertContent: "Welcome",
    isTimeUp: false,
    isTextBlinking: false,
    isTextRed: false,
  };

  handleSetMinute = (minute) => {
    this.resetCountdown(minute);
    this.terminateCountdown();

    setTimeout(() => {
      this.initializeCountdown();
    }, 600);
  };

  initializeCountdown = (speed = "1X", wasPaused = false) => {
    //set state and variables
    let state = this.state;
    let halfSecondExist = false;
    let halfMinute = 0;
    let timerSpeed = 1000;

    if (!wasPaused) {
      //check for half time
      if ((state.minute * 1) % 2 !== 0) halfSecondExist = true;

      if (state.minute === "1" || state.minute === 1) halfMinute = "00";
      else halfMinute = Math.floor((state.minute * 1) / 2);

      //setState
      this.setState({
        halfMinute,
        halfSecond: halfSecondExist ? "30" : "00",
      });
    }

    //check speed
    if (speed === "0.5X") {
      timerSpeed = 2000;
    } else if (speed === "-1X") {
      timerSpeed = 4000;
    }

    this.timer = setInterval(() => {
      //recall state each pass round
      state = this.state;

      //check for timeup
      if (
        (state.minute === "00" && state.second === "00") ||
        (state.minute * 1 <= 0 && state.second * 1 <= 0)
      ) {
        this.setTimedOut();
        this.terminateCountdown();
        return;
      }

      let minute = this.state.minute * 1;
      let second = this.state.second * 1;

      if (state.activeSpeed === "1.5X") {
        //check for less than 0 second
        if (second - 4 <= 0) {
          second = "58";
          minute -= 1;

          if (minute < 0) {
            this.setTimedOut();
            this.terminateCountdown();
            return;
          }
        } else {
          second -= 4;
        }
      } else if (state.activeSpeed === "2X") {
        //check for less than 0 second
        if (second - 6 <= 0) {
          second = "59";
          minute -= 1;

          if (minute < 0) {
            this.setTimedOut();
            this.terminateCountdown();
            return;
          }
        } else {
          second -= 6;
        }
      } else {
        //check for less than 0 second
        if (second <= 0) {
          second = "59";
          minute -= 1;
        } else {
          second -= 1;
        }
      }

      if (second < 10) {
        second = "0" + second;
      }

      if (minute < 10) {
        minute = "0" + minute;
      }

      //check for half time
      if (
        (minute * 1 <= state.halfMinute * 1 &&
          second * 1 <= state.halfSecond * 1) ||
        (state.halfSecond * 1 === 0 && minute * 1 < state.halfMinute * 1)
      ) {
        this.setState({
          isAlertOpen: true,
          alertContent: "More than halfway there!",
        });
      }

      //check for 20 seconds and less
      if (minute === "00" && second * 1 <= 20) {
        this.setState({ isTextRed: true });
      }

      //check for 10 seconds and less
      if (minute === "00" && second * 1 <= 10) {
        this.setState({ isTextBlinking: true });
      }

      this.setState({ minute, second });
    }, timerSpeed);
  };

  terminateCountdown = () => {
    clearInterval(this.timer);
  };

  setTimedOut = () => {
    this.setState({
      minute: "00",
      second: "00",
      isTimeUp: true,
      isAlertOpen: true,
      alertContent: "Time's Up!",
      isTextBlinking: false,
    });
  }

  resetCountdown = (minute = "00") => {
    this.setState({
      isAlertOpen: false,
      isTextRed: false,
      isTextBlinking: false,
      isTimeUp: false,
      activeControl: "None",
      activeSpeed: "1X",
      minute,
      second: "00",
    });
  };

  handleOnPauseClick = () => {
    this.setState({ clicked: "Pause", activeControl: "Paused" });

    this.terminateCountdown();

    setTimeout(() => {
      this.setState({
        clicked: "None",
        isAlertOpen: true,
        alertContent: "Timer Paused",
      });
    }, 1000);
  };

  handleOnPlayClick = () => {
    this.setState({ clicked: "Play", activeControl: "Played" });

    this.initializeCountdown(this.state.activeSpeed, true);

    setTimeout(() => {
      this.setState({ clicked: "None", isAlertOpen: false });
    }, 1000);
  };

  handleOnStopClick = () => {
    this.setState({ clicked: "Stop", activeControl: "Stoped" });

    this.terminateCountdown();
    this.resetCountdown();

    setTimeout(() => {
      this.setState({
        clicked: "None",
        isAlertOpen: true,
        alertContent: "Timer Stoped & cleared",
      });
    }, 1000);
  };

  handleOnSwitchSpeed = (speed) => {
    this.setState({ speedClicked: speed, activeSpeed: speed });

    this.terminateCountdown();
    this.initializeCountdown(speed, true);

    setTimeout(() => {
      this.setState({ speedClicked: "None" });
    }, 1000);
  };

  render() {
    return (
      <div className="Countdown">
        <Alert isOpen={this.state.isAlertOpen} isTimeUp={this.state.isTimeUp}>
          {this.state.alertContent}
        </Alert>

        {this.state.isTimeUp && (
          <audio src={CountdownOverSound} controls autoPlay hidden>
            <source src={CountdownOverSound} type="audio/mp3" />
            Your browser does not support the audio element.
          </audio>
        )}

        <Header />
        <div className="Countdown__body">
          <div className="Countdown__main">
            <div className="Countdown__main__left">
              <Controls
                icon={<Stop />}
                hoverIcon={<Stop hover={true} />}
                text="Stop/reset"
                textPos="left"
                controlClicked={this.state.clicked === "Stop"}
                onClick={() => this.handleOnStopClick()}
              />
            </div>
            <div className="Countdown__main__center">
              <Counter
                minute={this.state.minute}
                second={this.state.second}
                isPaused={this.state.activeControl === "Paused"}
                isTextRed={this.state.isTextRed}
                isTextBlinking={this.state.isTextBlinking}
                isTimeUp={this.state.isTimeUp}
              />
            </div>
            <div className="Countdown__main__right">
              {this.state.activeControl === "Paused" ? (
                <Controls
                  icon={<Play />}
                  hoverIcon={<Play hover={true} />}
                  text="Resume"
                  controlClicked={this.state.clicked === "Play"}
                  onClick={() => this.handleOnPlayClick()}
                />
              ) : (
                <Controls
                  icon={<Pause />}
                  hoverIcon={<Pause hover={true} />}
                  text="Pause"
                  controlClicked={this.state.clicked === "Pause"}
                  onClick={() => this.handleOnPauseClick()}
                />
              )}
            </div>
          </div>

          <div className="Countdown__speed">
            <h4>Speed Controls</h4>
            <div className="Countdown__speed__inner">
              <SpeedControls
                controlClicked={this.state.speedClicked === "1X"}
                active={this.state.activeSpeed === "1X"}
                onClick={() => this.handleOnSwitchSpeed("1X")}
              >
                1X
              </SpeedControls>

              <SpeedControls
                controlClicked={this.state.speedClicked === "1.5X"}
                active={this.state.activeSpeed === "1.5X"}
                onClick={() => this.handleOnSwitchSpeed("1.5X")}
              >
                1.5X
              </SpeedControls>

              <SpeedControls
                controlClicked={this.state.speedClicked === "2X"}
                active={this.state.activeSpeed === "2X"}
                onClick={() => this.handleOnSwitchSpeed("2X")}
              >
                2X
              </SpeedControls>

              <SpeedControls
                controlClicked={this.state.speedClicked === "0.5X"}
                active={this.state.activeSpeed === "0.5X"}
                onClick={() => this.handleOnSwitchSpeed("0.5X")}
              >
                0.5X
              </SpeedControls>

              <SpeedControls
                controlClicked={this.state.speedClicked === "-1X"}
                active={this.state.activeSpeed === "-1X"}
                onClick={() => this.handleOnSwitchSpeed("-1X")}
              >
                -1X
              </SpeedControls>
            </div>
          </div>
        </div>
        <Footer onStart={(minute) => this.handleSetMinute(minute)} />
      </div>
    );
  }
}

export default Countdown;
