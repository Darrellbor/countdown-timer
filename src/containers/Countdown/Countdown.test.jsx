import React from "react";
import { shallow } from "enzyme";

import Countdown from "./";

describe("Countdown Component", () => {
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = shallow(<Countdown />);
    instance = wrapper.instance();
  });

  test("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("handleSetMinute: should set minute state appropriately", () => {
    expect(wrapper.state("minute")).toBe("00");
    instance.handleSetMinute(5);
    expect(wrapper.state("minute")).toBe(5);
  });

  test("handleSetThemeColor: should set theme state appropriately", () => {
    expect(wrapper.state("theme")).toBe("Dark" || "Light");
    instance.handleSetThemeColor({
      target: { dataset: { setthemeto: "Light" } },
    });
    expect(wrapper.state("theme")).toBe("Light");
  });

  test("setTimedOut: should set state to time out appropriately", () => {
    instance.setTimedOut();
    expect(wrapper.state("minute")).toBe("00");
    expect(wrapper.state("second")).toBe("00");
    expect(wrapper.state("isTimeUp")).toBe(true);
    expect(wrapper.state("isAlertOpen")).toBe(true);
    expect(wrapper.state("alertContent")).toBe("Time's Up!");
    expect(wrapper.state("isTextBlinking")).toBe(false);
  });

  test("resetCountdown: should set state to default appropriately", () => {
    instance.resetCountdown();
    expect(wrapper.state("minute")).toBe("00");
    expect(wrapper.state("second")).toBe("00");
    expect(wrapper.state("isTimeUp")).toBe(false);
    expect(wrapper.state("isAlertOpen")).toBe(false);
    expect(wrapper.state("activeControl")).toBe("None");
    expect(wrapper.state("isTextBlinking")).toBe(false);
    expect(wrapper.state("activeSpeed")).toBe("1X");
  });

  test("handleOnPauseClick: should handle pause state with no time set", (done) => {
    instance.handleOnPauseClick();

    expect(wrapper.state("clicked")).toBe("Pause");

    setTimeout(() => {
      expect(wrapper.state("clicked")).toBe("None");
      done();
    }, 1000);
  });

  test("handleOnPauseClick: should handle pause state with time set", (done) => {
    instance.handleSetMinute(5);
    instance.handleOnPauseClick();

    expect(wrapper.state("clicked")).toBe("Pause");
    expect(wrapper.state("activeControl")).toBe("Paused");

    setTimeout(() => {
      expect(wrapper.state("clicked")).toBe("None");
      expect(wrapper.state("alertContent")).toBe("Timer Paused");
      expect(wrapper.state("isAlertOpen")).toBe(true);
      done();
    }, 1000);
  });

  test("handleOnStopClick: should handle stop state with no time set", (done) => {
    instance.handleOnStopClick();

    expect(wrapper.state("clicked")).toBe("Stop");

    setTimeout(() => {
      expect(wrapper.state("clicked")).toBe("None");
      done();
    }, 1000);
  });

  test("handleOnStopClick: should handle stop state with time set", (done) => {
    instance.handleSetMinute(5);
    instance.handleOnStopClick();

    expect(wrapper.state("clicked")).toBe("Stop");

    setTimeout(() => {
      expect(wrapper.state("clicked")).toBe("None");
      expect(wrapper.state("alertContent")).toBe("Timer Stoped & cleared");
      expect(wrapper.state("isAlertOpen")).toBe(true);
      done();
    }, 1000);
  });

  test("handleOnPlayClick: should handle on play appropriately", (done) => {
    instance.handleSetMinute(5);
    instance.handleOnPauseClick();
    instance.handleOnPlayClick();

    expect(wrapper.state("clicked")).toBe("Play");

    setTimeout(() => {
      expect(wrapper.state("activeControl")).toBe("Played");
      expect(wrapper.state("clicked")).toBe("None");
      expect(wrapper.state("isAlertOpen")).toBe(false);
      done();
    }, 1000);
  });

  test("handleOnSwitchSpeed: should handle on switch speed appropriately", (done) => {
    instance.handleSetMinute(5);
    instance.handleOnSwitchSpeed("2X");

    expect(wrapper.state("speedClicked")).toBe("2X");
    expect(wrapper.state("activeSpeed")).toBe("2X");

    setTimeout(() => {
      expect(wrapper.state("speedClicked")).toBe("None");
      done();
    }, 1000);
  });

  test("initializeCountdown: should update state if it wasn't paused and half time is calculated", () => {
    instance.handleSetMinute(5);
    instance.initializeCountdown();

    expect(wrapper.state("halfMinute")).toBe(2);
    expect(wrapper.state("halfSecond")).toBe("30");
  });

  test("initializeCountdown: should update second state every second", () => {
    instance.handleSetMinute(5);

    jest.useFakeTimers();
    expect(wrapper.state("minute")).toBe(5);
    instance.initializeCountdown();
    jest.advanceTimersByTime(3000);
    expect(wrapper.state("second")).toBe(57);
  });

  test("initializeCountdown: should update minute state every minute", () => {
    instance.handleSetMinute(7);

    jest.useFakeTimers();
    expect(wrapper.state("minute")).toBe(7);
    instance.initializeCountdown();
    jest.advanceTimersByTime(3000);
    expect(wrapper.state("minute")).toBe("06");
  });

  test("initializeCountdown: should update second state every 2s on 0.5x speed", () => {
    instance.handleSetMinute(7);

    jest.useFakeTimers();
    expect(wrapper.state("minute")).toBe(7);
    instance.initializeCountdown("0.5X");
    jest.runOnlyPendingTimers(2000);
    expect(wrapper.state("minute")).toBe("06");
    expect(wrapper.state("second")).toBe("59");
  });

  test("initializeCountdown: should update second state every 4s on -1x speed", () => {
    instance.handleSetMinute(9);

    jest.useFakeTimers();
    expect(wrapper.state("minute")).toBe(9);
    instance.initializeCountdown("-1X");
    jest.runOnlyPendingTimers(4000);
    expect(wrapper.state("minute")).toBe("08");
    expect(wrapper.state("second")).toBe("59");
  });

  test("terminateCountdown: should call clearinterval", () => {
    instance.handleSetMinute(9);

    jest.useFakeTimers();
    instance.terminateCountdown();
    expect(clearInterval).toHaveBeenCalledTimes(1);
  });
});
