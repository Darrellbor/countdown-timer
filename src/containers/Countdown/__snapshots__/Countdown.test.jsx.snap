// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown Component renders correctly 1`] = `
<div
  className="Countdown"
>
  <Alert
    isOpen={false}
    isTimeUp={false}
    theme="Dark"
  >
    Welcome
  </Alert>
  <ThemeController
    onClick={[Function]}
    theme="Dark"
  />
  <Header
    theme="Dark"
  />
  <div
    className="Countdown__body Countdown__body--dark"
  >
    <div
      className="Countdown__main"
    >
      <div
        className="Countdown__main__left"
      >
        <Controls
          controlClicked={false}
          hoverIcon={
            <Stop
              hover={true}
              theme="Dark"
            />
          }
          icon={
            <Stop
              hover={false}
              theme="Dark"
            />
          }
          onClick={[Function]}
          text="Stop/reset"
          textPos="left"
          theme="Dark"
        />
      </div>
      <div
        className="Countdown__main__center"
      >
        <Counter
          isPaused={false}
          isTextBlinking={false}
          isTextRed={false}
          isTimeUp={false}
          minute="00"
          second="00"
          theme="Dark"
        />
      </div>
      <div
        className="Countdown__main__right"
      >
        <Controls
          controlClicked={false}
          hoverIcon={
            <Pause
              hover={true}
              theme="Dark"
            />
          }
          icon={
            <Pause
              hover={false}
              theme="Dark"
            />
          }
          onClick={[Function]}
          text="Pause"
          textPos="right"
          theme="Dark"
        />
      </div>
    </div>
    <div
      className="Countdown__speed"
    >
      <h4>
        Speed Controls
      </h4>
      <div
        className="Countdown__speed__inner"
      >
        <SpeedControls
          active={true}
          controlClicked={false}
          onClick={[Function]}
          theme="Dark"
        >
          1X
        </SpeedControls>
        <SpeedControls
          active={false}
          controlClicked={false}
          onClick={[Function]}
          theme="Dark"
        >
          1.5X
        </SpeedControls>
        <SpeedControls
          active={false}
          controlClicked={false}
          onClick={[Function]}
          theme="Dark"
        >
          2X
        </SpeedControls>
        <SpeedControls
          active={false}
          controlClicked={false}
          onClick={[Function]}
          theme="Dark"
        >
          0.5X
        </SpeedControls>
        <SpeedControls
          active={false}
          controlClicked={false}
          onClick={[Function]}
          theme="Dark"
        >
          - 1X
        </SpeedControls>
      </div>
    </div>
  </div>
  <Footer
    onStart={[Function]}
    theme="Dark"
  />
</div>
`;
